### ISSUE TITLE
Crash on Http2ConnectionState::release_stream()

### ISSUE THREAD

**Issue by masaori335** [apache/trafficserver#4504]
We got this crash on 8.0.x. This looks pretty similar to #4158. So #4158 (adding null check) could be workaround fix. But it looks like there is a race on shutdown which we should fix.

```
[ 0  ] libc-2.12.so       __libc_waitpid                                               ( :undefined                   ) 
[ 1  ] traffic_server     crash_logger_invoke                                          ( Crash.cc:165                 ) 
[ 2  ] libc-2.12.so       0x2b0b38dfd570                                               ( :undefined                   ) 
[ 3  ] traffic_server     Http2ConnectionState::release_stream(Http2Stream*)           ( Http2ConnectionState.cc:1261 ) 
[ 4  ] traffic_server     Http2Stream::destroy()                                       ( Http2Stream.cc:730           ) 
[ 5  ] traffic_server     Http2Stream::terminate_if_possible()                         ( Http2Stream.cc:380           ) 
[ 6  ] traffic_server     Http2Stream::main_event_handler(int, void*)                  ( Http2Stream.cc:131           ) 
[ 7  ] traffic_server     handleEvent                                                  ( I_Continuation.h:160         ) 
[ 8  ] traffic_server     EThread::process_event(Event*, int)                          ( UnixEThread.cc:131           ) 
[ 9  ] traffic_server     EThread::process_queue(Queue<Event, Event::Link_link>*, i... ( UnixEThread.cc:170           ) 
[ 10 ] traffic_server     EThread::execute_regular()                                   ( UnixEThread.cc:230           ) 
[ 11 ] traffic_server     spawn_thread_internal                                        ( Thread.cc:85                 ) 
[ 12 ] libpthread-2.12.so start_thread                                                 ( :undefined                   )
```

https://github.com/apache/trafficserver/blob/8c50c75ac5faccbc2aa7d1b3880a7c4205ad8c96/proxy/http2/Http2ConnectionState.cc#L1257-L1265

**Comment by zwoop** [comment]
Here's a more detailed stack trace:

```
(gdb) bt full
#0  0x00002b0b38e778dd in waitpid () from /lib64/libc.so.6
No symbol table info available.
#1  0x0000559a4e823a94 in crash_logger_invoke (signo=11, info=0x2b0b4130b730, ctx=0x2b0b4130b600) at traffic_server/Crash.cc:165
        status = 0
#2  <signal handler called>
No symbol table info available.
#3  0x0000559a4e8eebbd in Http2ConnectionState::release_stream (this=0x2b462a029cb0, stream=<optimized out>) at Http2ConnectionState.cc:1261
        lock = {m = {m_ptr = 0x2b3dd1f97950}, locked_p = true}
#4  0x0000559a4e8e2975 in Http2Stream::destroy (this=0x2b4a06bee400) at Http2Stream.cc:730
        h2_parent = <optimized out>
        lock = <optimized out>
        end_time = <optimized out>
        lock = <optimized out>
#5  0x0000559a4e8e06d6 in Http2Stream::terminate_if_possible (this=0x2b4a06bee400) at Http2Stream.cc:380
        h2_parent = <optimized out>
        lock = <optimized out>
#6  0x0000559a4e8df306 in Http2Stream::main_event_handler (this=0x2b4a06bee400, event=103, edata=<optimized out>) at Http2Stream.cc:131
        e = <optimized out>
        this_thread = <optimized out>
        lock = <optimized out>
#7  0x0000559a4ea3ad80 in Continuation::handleEvent (this=<optimized out>, event=<optimized out>, data=<optimized out>) at ./I_Continuation.h:160
No locals.
#8  EThread::process_event (this=0x2b0b3cc0d1c0, e=0x2b47449e6480, calling_code=103) at UnixEThread.cc:131
        c_temp = <optimized out>
        lock = <optimized out>
#9  0x0000559a4ea3b01b in EThread::process_queue (this=0x2b0b3cc0d1c0, NegativeQueue=0x2b0b4130bdc0, ev_count=0x2b0b4130bde4, nq_count=0x2b0b4130bde0) at UnixEThread.cc:170
No locals.
#10 0x0000559a4ea3b4cb in EThread::execute_regular (this=0x2b0b3cc0d1c0) at UnixEThread.cc:230
        sleep_time = <optimized out>
        done_one = <optimized out>
        METRIC_INIT = {_loop_time = {_start = 0, _min = 9223372036854775807, _max = 0}, _events = {_min = 2147483647, _max = 0, _total = 0}, _count = 0, _wait = 0}
        delta = 0
        next_time = 0
        nq_count = 5399
        prev_metric = 0x2b0b3cd115c0
        ev_count = 5400
        loop_start_time = <optimized out>
        loop_finish_time = <optimized out>
#11 0x0000559a4ea3a3ec in spawn_thread_internal (a=0x2b0b3a154d80) at Thread.cc:85
        p = 0x2b0b3a154d80
#12 0x00002b0b39166aa1 in start_thread () from /lib64/libpthread.so.0
No symbol table info available.
#13 0x00002b0b38eb3c4d in clone () from /lib64/libc.so.6
No symbol table info available.
```

**Comment by zwoop** [comment]
I think this was fixed in a639581c85d9b9c9dcb728b03e86efc280b303c3. I'm going to close this.

**Comment by masaori335** [comment]
Unfortunately, we and @zizhong still see this crash with a639581.

**Comment by zwoop** [comment]
Yeh, the crasher is less frequent, but still happens. Seems we just reduce the window of the race condition?

**Comment by zwoop** [comment]
Masaori added a "history" feature, which I'm including here:

```
$2 = {history = {{location = {file = 0x556774d56dd8 "Http2ConnectionState.cc", func = 0x556774d56ff0 "delete_stream", line = 1241}, event = 34463, reentrancy = 1}, {location = {
        file = 0x556774d56dd8 "Http2ConnectionState.cc", func = 0x556774d5698c "main_event_handler", line = 915}, event = 34463, reentrancy = 1}, {location = {
        file = 0x556774d56dd8 "Http2ConnectionState.cc", func = 0x556774d5702f "release_stream", line = 1272}, event = 34463, reentrancy = 1}, {location = {
        file = 0x556774d56dd8 "Http2ConnectionState.cc", func = 0x556774d5702f "release_stream", line = 1272}, event = 34463, reentrancy = 0}, {location = {
        file = 0x556774d56dd8 "Http2ConnectionState.cc", func = 0x556774d5702f "release_stream", line = 1272}, event = 34463, reentrancy = 0}, {location = {file = 0x0,
        func = 0x0, line = 0}, event = 0, reentrancy = 0} <repeats 60 times>}, history_pos = 5}
```

**Comment by masaori335** [comment]
FWIW, this is history of Http2ClientSession
```
(gdb) frame 4
#4  0x000056375b2e394e in Http2Stream::destroy (this=0x7f89e383cbe0) at Http2Stream.cc:734
(gdb) p parent
$3 = (ProxyClientSession *) 0x7fc257fd3080
(gdb) p ((Http2ClientSession *)0x7fc257fd3080)->_history
$5 = {history = {{location = {file = 0x56375b46664c "Http2ClientSession.cc", func = 0x56375b46f450 "start", line = 163}, event = 34463, reentrancy = 0}, {
      location = {file = 0x56375b46664c "Http2ClientSession.cc", func = 0x56375b4668d3 "state_read_connection_preface", line = 414}, event = 100,
      reentrancy = 1}, {location = {file = 0x56375b46664c "Http2ClientSession.cc", func = 0x56375b4668d3 "state_read_connection_preface", line = 414},
      event = 100, reentrancy = 1}, {location = {file = 0x56375b46664c "Http2ClientSession.cc", func = 0x56375b4668d3 "state_read_connection_preface",
        line = 432}, event = 34463, reentrancy = 1}, {location = {file = 0x56375b46664c "Http2ClientSession.cc",
        func = 0x56375b466970 "state_start_frame_read", line = 459}, event = 100, reentrancy = 2}, {location = {
        file = 0x56375b46664c "Http2ClientSession.cc", func = 0x56375b466af2 "do_complete_frame_read", line = 535}, event = 34463, reentrancy = 2}, {
      location = {file = 0x56375b46664c "Http2ClientSession.cc", func = 0x56375b466af2 "do_complete_frame_read", line = 535}, event = 34463,
      reentrancy = 2}, {location = {file = 0x56375b46664c "Http2ClientSession.cc", func = 0x56375b466af2 "do_complete_frame_read", line = 535},
      event = 34463, reentrancy = 2}, {location = {file = 0x56375b46664c "Http2ClientSession.cc", func = 0x56375b466af2 "do_complete_frame_read",
        line = 535}, event = 34463, reentrancy = 2}, {location = {file = 0x56375b46664c "Http2ClientSession.cc", func = 0x56375b4504a8 "destroy",
        line = 77}, event = 34463, reentrancy = 1}, {location = {file = 0x56375b46664c "Http2ClientSession.cc", func = 0x56375b483c7a "do_io_close",
        line = 297}, event = 34463, reentrancy = 1}, {location = {file = 0x56375b46664c "Http2ClientSession.cc",
        func = 0x56375b46674c "main_event_handler", line = 403}, event = 105, reentrancy = 0}, {location = {file = 0x56375b46664c "Http2ClientSession.cc",
        func = 0x56375b466679 "free", line = 145}, event = 34463, reentrancy = 0}, {location = {file = 0x0, func = 0x0, line = 0}, event = 0,
      reentrancy = 0} <repeats 52 times>}, history_pos = 13}
```

**Comment by masaori335** [comment]
The history gave us clue! Masakazu and I found below.

### TLDR; 
Calling `vc->add_to_keep_alive_queue()` (line 1253) made `ua_session` `nullptr` with heavy traffic.

https://github.com/apache/trafficserver/blob/ce706ce7be76bc08a49b86d58c08f01ba55684bf/proxy/http2/Http2ConnectionState.cc#L1246-L1261

### Details
It looks like below was happened

1. `Http2ConnectionState::release_stream()` did null check of `ua_session`
2. `UnixNetVConnection::add_to_keep_alive_queue()` is called
3. `NetHandler::manage_keep_alive_queue()` is called
4. If there is many inactive connection, `NetHandler::manage_keep_alive_queue()` start closing inactive connections with `VC_EVENT_INACTIVITY_TIMEOUT (105)`.
5. The event comes `Http2ClientSession::main_event_handler()`.
6. `Http2ClientSession::main_event_handler()` free itself.
7. `Http2ClientSession::free()` calls `Http2ConnectionState::destroy()`
8. `ua_session` became `nullptr`

### Reproduction
I could reproduced similar crash in local box with small `proxy.config.net.max_connections_in` and h2load. The call stack which I could reproduce is https://paste.apache.org/hd4e

### Conditions 
This crash could happen only if `total_connections_in > max_connections_per_thread_in` in `NetHandler::manage_keep_alive_queue()`.

https://github.com/apache/trafficserver/blob/ce706ce7be76bc08a49b86d58c08f01ba55684bf/iocore/net/UnixNet.cc#L610-L612

From the core file, when ATS crashed, the condition was satisfied.

```
(gdb) frame 9
#9  0x000056375b43d10b in EThread::process_queue (this=0x7fc39120d740, NegativeQueue=0x7fc38b70bdc0, ev_count=0x7fc38b70bde4, nq_count=0x7fc38b70bde0) at UnixEThread.cc:170
170	      process_event(e, e->callback_event);
(gdb) p this
$2 = (EThread *) 0x7fc39120d740
(gdb) p unix_netProcessor.netHandler_offset
$3 = 15584
(gdb) p (NetHandler *)((char *)(0x7fc39120d740) + 15584)
$4 = (NetHandler *) 0x7fc391211420
(gdb) p ((NetHandler *)((char *)(0x7fc39120d740) + 15584))->active_queue_size
$5 = 692
(gdb) p ((NetHandler *)((char *)(0x7fc39120d740) + 15584))->keep_alive_queue_size
$6 = 9118
(gdb) p ((NetHandler *)((char *)(0x7fc39120d740) + 15584))->max_connections_per_thread_in
$7 = 3125
(gdb) p ((NetHandler *)((char *)(0x7fc39120d740) + 15584))->thread
$8 = (EThread *) 0x7fc39120d740
```

I'm going to open PR to avoid this.
BTW, the @dmorilha's first patch (adding null check) was correct in some way :p

**Comment by zizhong** [comment]
Nicely done! @masaori335

**Comment by dmorilha** [comment]
nice! are we back porting this to 7.1 ?

**Comment by zwoop** [comment]
Yeh, seems like a reasonable candidate for 7.1.x. Is anyone testing H2 on 7.1 ?
